File: core/src/test/scala/diplomacy/variant/standard/rule/RetreatOrderGeneratorSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map.Power
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board.Keywords._
import diplomacy.variant.standard.board.Implicits._

class RetreatOrderGeneratorSpec extends UnitSpec {
  val generator = new RetreatOrderGenerator[T, map.Power]
  val board = {
    new Board[T, map.Power](
      map.map,
      1901.Spring - Retreat,
      Set(
        DiplomacyUnit(France, Army, Bur),
        DiplomacyUnit(France, Army, Mar),
        DiplomacyUnit(Italy, Fleet, Wes)
      ), Map(),
      Map(
        DiplomacyUnit(France, Army, Mar) -> UnitStatus.Dislodged(Gas.province),
        DiplomacyUnit(Italy, Fleet, Wes) -> UnitStatus.Dislodged(Tyn.province)
      ),
      Map(Pie.province -> ProvinceStatus.Standoff)
    )
  }

  "A movement-order-generator" should {
    "use Disband as a default order." in {
      val $ = StandardRuleOrderHelper(board)
      import $._
      generator.defaultOrder(board)(A(Mar)) should be(A(Mar).disband())
      generator.defaultOrder(board)(F(Wes)) should be(F(Wes).disband())
    }
  }
  it when {
    "all dislodged units cannot retreat" should {
      "use disband orders to skip the retreat phase." in {
        val board = {
          new Board[T, map.Power](
            map.map, 1901.Spring - Retreat,
            Set(
              DiplomacyUnit(Germany, Army, Par),
              DiplomacyUnit(Germany, Army, Gas),
              DiplomacyUnit(Germany, Army, Bre),
              DiplomacyUnit(France, Army, Bre)
            ),
            Map(), Map(DiplomacyUnit(France, Army, Bre) -> UnitStatus.Dislodged(Pic.province)),
            Map()
          )
        }
        val $ = StandardRuleOrderHelper(board)
        import $._

        generator.ordersToSkipPhase(board) should be(Some(Set(A(Bre).disband())))
      }
    }

    "there are some units that can retreat" can {
      "not skip the retreat phase." in {
        generator.ordersToSkipPhase(board) should be(None)
      }
    }
  }
}

File: test/variant/standard/rule/retreat-order-generator-spec.js
'use strict'

const chai = require('chai')

const { Board, Unit } = require('../../../../lib/board/package')
const { map, location, Power } = require('../../../../lib/variant/standard/map/package')
const { Turn, Season } = require('../../../../lib/variant/standard/board/package')
const { Army, Fleet } = require('../../../../lib/variant/standard/rule/military-branch')
const { Retreat } = require('../../../../lib/variant/standard/rule/phase')
const Dislodged = require('../../../../lib/variant/standard/rule/dislodged')
const ProvinceStatus = require('../../../../lib/variant/standard/rule/province-status')
const State = require('../../../../lib/variant/standard/rule/state')
const RetreatOrderGenerator = require('../../../../lib/variant/standard/rule/retreat-order-generator')
const Helper = require('../../../../lib/variant/standard/rule/standard-rule-helper')

chai.should()

const $ = location
const generator = new RetreatOrderGenerator()

describe('A RetreatOrderGenerator', () => {
  it('uses Disband as a default order', () => {
    const u1 = new Unit(Army, $.Mar, Power.France)
    const u2 = new Unit(Fleet, $.Wes, Power.Italy)
    const board = new Board(
      map, new State(new Turn(1901, Season.Spring), Retreat),
      [
        new Unit(Army, $.Bur, Power.France),
        u1,
        u2
      ],
      [
        [u1, new Dislodged($.Gas.province)], [u2, new Dislodged($.Tyn.province)]
      ],
      [[$.Pie.province, new ProvinceStatus(null, true)]]
    )
    const $$ = new Helper(board)
    generator.defaultOrderOf(board, $$.A($.Mar).unit).should.deep.equal($$.A($.Mar).disband())
    generator.defaultOrderOf(board, $$.F($.Wes).unit).should.deep.equal($$.F($.Wes).disband())
  })
  describe('when all dislodged units cannot retreat', () => {
    it('uses disband orders to skip the retreat phase', () => {
      const u1 = new Unit(Army, $.Bre, Power.France)
      const board = new Board(
        map, new State(new Turn(1901, Season.Spring), Retreat),
        [
          new Unit(Army, $.Par, Power.Germany),
          new Unit(Army, $.Gas, Power.Germany),
          new Unit(Army, $.Bre, Power.Germany),
          u1
        ],
        [[u1, new Dislodged($.Pic.province)]],
        []
      )
      const $$ = new Helper(board)
      generator.ordersToSkipPhase(board).should.deep.equal(new Set([$$.A($.Bre).disband()]))
    })
  })
  describe('there  are some units that can retreat', () => {
    it('does not skip the retreat phase', () => {
      const u1 = new Unit(Army, $.Mar, Power.France)
      const u2 = new Unit(Fleet, $.Wes, Power.Italy)
      const board = new Board(
        map, new State(new Turn(1901, Season.Spring), Retreat),
        [
          new Unit(Army, $.Bur, Power.France),
          u1,
          u2
        ],
        [
          [u1, new Dislodged($.Gas.province)], [u2, new Dislodged($.Tyn.province)]
        ],
        [[$.Pie.province, new ProvinceStatus(null, true)]]
      );
      (generator.ordersToSkipPhase(board) === null).should.equal(true)
    })
  })
})
